feat(Client): add max response size option

mateonunez opened the pull request:
This PR implements #1692 by introducing a new client option: `maxResponseSize`.

The `maxResponseSize` parameter is set to `0` which means whatever size is allowed, otherwise, the size will be considered in bytes.

There is a new check before the _Parser_ computes the `bytesRead`. If that computation is greater than the `maxResponseSize` the socket is destroyed by the corresponding error.

```js
    if (maxResponseSize !== 0 && this.bytesRead + buf.length > maxResponseSize) {
      util.destroy(socket, new ResponseExceededMaxSizeError())
      return -1
    }
```

This approach enforces the `Client` to verify large responses and to close the socket before the buffer is completed.

This implementation surf in a low-level way, so tests (such as _ResponseContentLengthMismatchError_) are covered by `scenarios` in `test/errors.js`

mcollina reviewed (changes requested):
Thanks for opening a PR! Can you please add a unit test?

KhafraDev commented:
Should 0 be the default - is there ever a possibility that someone would want to ensure a response has an empty body? If so, -1 would be a better default for unlimited body size.

mateonunez commented:
> Should 0 be the default - is there ever a possibility that someone would want to ensure a response has an empty body? If so, -1 would be a better default for unlimited body size.

I fully agree with you, I changed the default value to **-1** and added a test to check even an empty payload.

KhafraDev commented on test/max-response-size.js:
```diff
@@ -0,0 +1,103 @@
+    t.plan(2)
+
+    t.throws(() => {
+      Client('http://localhost:3000', { maxResponseSize: 'hello' })
```
```suggestion
      new Client('http://localhost:3000', { maxResponseSize: 'hello' })
```

mateonunez replied:
In this way I got a `standard` error:

```bash
❯ npm run lint

> undici@5.11.0 lint
> standard | snazzy

standard: Use JavaScript Standard Style (https://standardjs.com)

/Users/mateonunez/dev/undici/test/max-response-size.js
   97:7  error  Do not use 'new' for side effects  no-new
  100:7  error  Do not use 'new' for side effects  no-new

✖ 2 problems
```

There is a way to avoid it without disabling the lint rule?

ronag replied:
Assign the object to a variable or use eslint-disable-line

KhafraDev commented on test/max-response-size.js:
```diff
@@ -0,0 +1,103 @@
+      Client('http://localhost:3000', { maxResponseSize: 'hello' })
+    }, 'maxResponseSize must be a number')
+    t.throws(() => {
+      Client('http://localhost:3000', { maxResponseSize: -2 })
```
```suggestion
      new Client('http://localhost:3000', { maxResponseSize: -2 })
```

ronag commented on lib/client.js:
```diff
@@ -209,6 +216,7 @@ class Client extends DispatcherBase {
     this[kMaxRedirections] = maxRedirections
     this[kMaxRequests] = maxRequestsPerClient
     this[kClosedResolve] = null
+    this[kMaxResponseSize] = maxResponseSize || -1
```
This won't allow 0

mateonunez commented:
Thanks for the review @ronag and @KhafraDev 🙏. I made the requested changes, disabling the lint rule and fixing the `maxResponseSize` assignment.

nosovk commented:
@mateonunez can you please add `maxResponseSize` to fetch options also, to close https://github.com/nodejs/undici/issues/1692?

KhafraDev commented:
That should not be added to fetch options
